processor/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+            pc: 0,
+            flag: false,
+            memory: Memory{state: HashMap::new(), trace: BTreeMap::new()}
+        };
+    }
+
+    pub fn get_reg_index(&self, reg_str: &str) -> usize {
+        let first = reg_str.chars().nth(0).unwrap();
+        assert!(first ==  'r', "wrong reg name");
+        let index = reg_str[1..].parse().unwrap();
+        return index;
+    }
+
+    pub fn get_index_value(&self, op_str: &str)-> (GoldilocksField, ImmediateOrRegName) {
+        let src = op_str.parse();
+        let mut value = Default::default();
+        if src.is_ok() {
+            let data: u64 = src.unwrap();
+            return (GoldilocksField::from_canonical_u64(data), ImmediateOrRegName::Immediate(GoldilocksField::from_canonical_u64(data)));
+        } else {
+            let src_index = self.get_reg_index(op_str);
+            value = self.registers[src_index];
+            return (value, ImmediateOrRegName::RegName(src_index as u8));
+        }
+    }
+
+    pub fn decode_instruction(&self, raw_instruction: String) -> Instruction {
+        let mut ops: Vec<_> = raw_instruction.split(' ').collect();
+        let mut opcode = ops.get(0).unwrap().to_lowercase();
+
+        let instuction = match opcode.as_str() {
+            "mov" => {
+                debug!("opcode: mov");
+                assert!(ops.len() == 3 ,  "mov params len is 2");
+                let dst_index = self.get_reg_index(&ops[1]);
+                let value = self.get_index_value(&ops[2]);
+                Instruction::MOV(Mov{ri:dst_index as u8,a:value.1})
+
+            },
+            "eq" => {
+                debug!("opcode: eq");
+                assert!(ops.len() == 3 ,  "eq params len is 2");
+                let dst_index = self.get_reg_index(&ops[1]);
+                // let src_index = self.get_reg_index(&ops[2]);
+                let value = self.get_index_value(&ops[2]);
+                Instruction::EQ(Equal{ri: dst_index as u8, a: value.1})
+            }
+            "cjmp" => {
+                debug!("opcode: cjmp");
+                assert!(ops.len() == 2 ,  "cjmp params len is 1");
+                let value = self.get_index_value(&ops[1]);
+                Instruction::CJMP(CJmp{a: value.1})
+            }
+            "jmp" => {
+                debug!("opcode: jmp");
+                assert!(ops.len() == 2 ,  "jmp params len is 1");
+                let value = self.get_index_value(&ops[1]);
+                Instruction::JMP(Jmp{a: value.1})
+
+            }
+            "add" => {
+                debug!("opcode: add");
+                assert!(ops.len() == 4 ,  "add params len is 3");
+                let dst_index = self.get_reg_index(&ops[1]);
+                let op1_index = self.get_reg_index(&ops[2]);
+                let op2_value = self.get_index_value(&ops[3]);
+                Instruction::ADD(Add{ri: dst_index as u8, rj: op1_index as u8, a: op2_value.1})
+            }
+            "sub" => {
```

Comment:
why do we have sub-instruction? For consistency on design , we need remove it like div-instruction